cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/NamespaceSourceControlMetadataRefresher.java

```diff
@@ -0,0 +1,194 @@
+
+    Set<String> repoFileNames = repositoryAppsResponse.getApps().stream()
+        .map(RepositoryApp::getName)
+        .collect(Collectors.toCollection(HashSet::new));
+
+    TransactionRunners.run(transactionRunner, context -> {
+      RepositorySourceControlMetadataStore repoMetadataStore = getRepoSourceControlMetadataStore(
+          context);
+
+      // Cleaning up the repo source control metadata table
+      cleanupRepoSourceControlMeta(namespaceId.getNamespace(),
+          repoFileNames, repoMetadataStore);
+
+      // Updating the namespace and repo source control metadata table
+      NamespaceSourceControlMetadataStore namespaceMetadataStore =
+          getNamespaceSourceControlMetadataStore(context);
+      for (RepositoryApp repoApp : repositoryAppsResponse.getApps()) {
+        updateSourceControlMeta(
+            new ApplicationReference(namespaceId.getNamespace(), repoApp.getName()),
+            repoApp.getFileHash(),
+            refreshStartTime, repoMetadataStore, namespaceMetadataStore);
+      }
+    });
+    lastRefreshTime.set(refreshStartTime);
+    return this.lastRefreshTime.get();
+  }
+
+  private void updateSourceControlMeta(ApplicationReference appRef,
+      String repoFileHash,
+      long refreshStartTime, RepositorySourceControlMetadataStore repoMetadataStore,
+      NamespaceSourceControlMetadataStore namespaceMetadataStore) throws IOException {
+
+    SourceControlMeta namespaceSourceControlMeta = namespaceMetadataStore.get(appRef);
+
+    if (namespaceSourceControlMeta == null) {
+      repoMetadataStore.write(appRef, false, 0L);
+      return;
+    }
+
+    if (namespaceSourceControlMeta.getLastSyncedAt() != null
+        && namespaceSourceControlMeta.getLastSyncedAt().isAfter(
+        Instant.ofEpochMilli(refreshStartTime))) {
+      repoMetadataStore.write(appRef,
+          Boolean.TRUE.equals(namespaceSourceControlMeta.getSyncStatus()),
+          namespaceSourceControlMeta.getLastSyncedAt().toEpochMilli());
+      return;
+    }
+    boolean isSynced = repoFileHash.equals(namespaceSourceControlMeta.getFileHash());
+    namespaceMetadataStore.write(appRef,
+        SourceControlMeta.builder(namespaceSourceControlMeta).setSyncStatus(isSynced).build());
+    repoMetadataStore.write(appRef,
+        isSynced, namespaceSourceControlMeta.getLastSyncedAt() == null ? 0L :
+            namespaceSourceControlMeta.getLastSyncedAt().toEpochMilli());
+  }
+
+  private void cleanupRepoSourceControlMeta(String namespace,
+      Set<String> repoFiles,
+      RepositorySourceControlMetadataStore store) throws IOException {
+    ScanSourceControlMetadataRequest request = ScanSourceControlMetadataRequest
+        .builder()
+        .setNamespace(namespace)
+        .setLimit(Integer.MAX_VALUE)
+        .build();
+    List<SourceControlMetadataRecord> records = new ArrayList<>();
+    String type = EntityType.APPLICATION.toString();
+
+    store.scan(request, type, records::add);
```

Comment:
you can do 
```
if (!repoFiles.contains(record.getName())) {
        store.delete(new ApplicationReference(record.getNamespace(), record.getName()));
      }
```
inside scan

Reply:
This creates a nested transaction since you are calling this method from inside a tx runner. Please don't do it.